stellar-multitenant/internal/services/status_change_validator.go

```diff
@@ -0,0 +1,65 @@
+package services
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/stellar/go/support/log"
+
+	"github.com/stellar/stellar-disbursement-platform-backend/internal/data"
+	"github.com/stellar/stellar-disbursement-platform-backend/stellar-multitenant/pkg/tenant"
+)
+
+var (
+	ErrCannotRetrieveTenantByID  = errors.New("cannot retrieve tenant by id")
+	ErrCannotRetrievePayments    = errors.New("cannot retrieve payments for tenant")
+	ErrCannotDeactivateTenant    = errors.New("cannot deactivate tenant")
+	ErrCannotActivateTenant      = errors.New("cannot activate tenant for tenant that is not deactivated")
+	ErrCannotPerformStatusUpdate = errors.New("cannot perform update on tenant to requested status")
+)
+
+func ValidateStatus(ctx context.Context, manager tenant.ManagerInterface, models *data.Models, tenantID string, reqStatus tenant.TenantStatus) error {
+	tnt, err := manager.GetTenant(ctx,
+		&tenant.QueryParams{
+			Filters: map[tenant.FilterKey]interface{}{
+				tenant.FilterKeyID: tenantID,
+			},
+		})
+	if err != nil {
+		return fmt.Errorf("%w: %w", ErrCannotRetrieveTenantByID, err)
+	}
+	ctx = tenant.SaveTenantInContext(ctx, tnt)
+
+	// if attempting to deactivate tenant, need to check for a few conditions such as
+	// 1. whether tenant is already deactivated
+	// 2. whether there are any payments not in a terminal state
+	if reqStatus == tenant.DeactivatedTenantStatus {
+		if tnt.Status == tenant.DeactivatedTenantStatus {
+			log.Ctx(ctx).Warnf("tenant %s is already deactivated", tenantID)
+		} else {
+			indeterminatePaymentsCount, getPaymentCountErr := models.Payment.Count(ctx, &data.QueryParams{
+				Filters: map[data.FilterKey]interface{}{
+					data.FilterKeyStatus: data.PaymentNonTerminalStatuses(),
```

Comment:
I see that we avoid deactivating an instance that has `NonTerminalStatus`. 
`DraftPaymentStatus, ReadyPaymentStatus, PendingPaymentStatus, PausedPaymentStatus`

I think it's ok to deactivate an instance with `DraftPaymentStatus` and potentially even `PausedPaymentStatus`. Thoughts on this @marcelosalloum ?

Reply:
I agree! @ziyliu: I think you can replace the method `data.PaymentNonTerminalStatuses()` with something that denotes "active/inProgress/pending payments".